### PR TITLE
unix: Set a guard page at the end of signal stacks

### LINKED ISSUE #69533: Rust's signal stack should be guarded against overflow
Rust's runtime installs a `SIGSEGV` signal handler to distinguish stack overflows from other faults. This requires using an alternate signal stack (`sigaltstack`), which we create with the default size `SIGSTKSZ`. That's usually quite small, 8KB or so depending on the target, but it's still plenty for what the `SIGSEGV` handler is doing. There's currently no guard page like we have on normal stacks.

That alternate stack is shared by all signals for a thread, so if a user installs any other signal handler, and they opt in with the flag `SA_ONSTACK`, they'll be limited to the same tiny stack. If their handler happens to be more involved and overflows that stack, it may silently clobber other memory. If we had a guard page on the signal stack, that would cause a process abort instead.

This was reported to the Security Response WG with a reproducer in [this repo](https://github.com/eust/rust_sigstack_overflow). However, after discussion we decided to treat this as a normal issue for a few reasons:

- Setting a custom signal handler requires `unsafe` code.
- Signal handlers already run in a quite constrained environment, e.g. [`signal-safety(7)`](http://man7.org/linux/man-pages/man7/signal-safety.7.html).
- It's unusual to use `SA_ONSTACK` for signals other than `SIGSEGV`, and if you choose so, you also bear responsibility to make sure that stack suffices.

Nevertheless, as a defensive measure, it's still a reasonable idea to map a `PROT_NONE` guard page when Rust is creating its signal stack, so an unexpected overflow will abort rather than silently corrupting memory. For full protection, that would also depend on the guard page not being skipped, but not all targets have stack probes yet (#43241).

---

Thanks to @eust for reporting this in accordance with our [security policy](https://www.rust-lang.org/policies/security), even though we have decided not to treat this as a security issue at this time.

cc @rust-lang/security

### PR DESCRIPTION
**Pull request by iximeow**
This mitigates possible issues when signal stacks overflow, which could
manifest as segfaults or in unlucky circumstances possible clobbering of
other memory values as stack overflows tend to enable.

I went ahead and made a PR for this because it's a pretty small change, though if I should open an issue/RFC for this and discuss there first I'll happily do so. I've also added some example programs that demonstrate the uncomfortably clobber-happy behavior we currently have, and the segfaults that could/should result instead, [here](https://github.com/iximeow/jubilant-train).

### CODE REVIEW

**Comment by 9999years** [comment]
> #69969

Nice.

**Comment by cuviper** [comment]
This should close #69533.

File: src/libstd/sys/unix/stack_overflow.rs [hunk 1]
```diff
@@ -137,11 +137,16 @@ mod imp {
...
-            mmap(ptr::null_mut(), SIGSTKSZ, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+            mmap(ptr::null_mut(), SIGSTKSZ + page_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
         if stackp == MAP_FAILED {
             panic!("failed to allocate an alternative stack");
         }
+        let guard_result = libc::mprotect(stackp, page_size, PROT_NONE);
+        if guard_result != 0 {
+            panic!("failed to set up alternative stack guard page");
+        }
         stackp
```
Comment (cuviper) [inline src/libstd/sys/unix/stack_overflow.rs]: You're returning the protected page as the bottom of the stack. That will sort of work as long as `SIGSTKSZ` is larger than the page size to begin with, so the top will still be in R/W memory, but I think you want to return `stackp.add(page_size)`.

The `drop_handler` also needs to be adjusted to account for the correct map size.

**Review by cuviper — CHANGES_REQUESTED** [review]

File: src/libstd/sys/unix/stack_overflow.rs (hunk 1 above)
Comment (iximeow) [reply src/libstd/sys/unix/stack_overflow.rs]: I'd entirely neglected `drop_handler`! Thank you. I've adjusted `get_stackp` to return `stackp.add(page_size)` as well.

File: src/libstd/sys/unix/stack_overflow.rs [hunk 2]
```diff
@@ -138,11 +139,15 @@ mod imp {
 
     unsafe fn get_stackp() -> *mut libc::c_void {
         let stackp =
-            mmap(ptr::null_mut(), SIGSTKSZ, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+            mmap(ptr::null_mut(), SIGSTKSZ + page_size(), PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
```
Comment (cuviper) [inline src/libstd/sys/unix/stack_overflow.rs]: This long line is going to get flagged by `./x.py fmt --check`. You could compute the size separately, or just accept the reflow by running `./x.py fmt` yourself.

**Review by cuviper — APPROVED** [review]
LGTM, but someone from @rust-lang/libs should probably give final say.

**Comment by cuviper** [comment]
> I've also added some example programs that demonstrate the uncomfortably clobber-happy behavior we currently have, and the segfaults that could/should result instead, [here](https://github.com/iximeow/jubilant-train).

How reliably are these tests able to get a nearby mapping to clobber? It would be nice to have a regression test, something like `src/test/ui/out-of-stack.rs` that watches for `SIGABRT`.

File: src/libstd/sys/unix/stack_overflow.rs (hunk 2 above)
Comment (iximeow) [reply src/libstd/sys/unix/stack_overflow.rs]: I was wondering about that. I naively tried a `cargo fmt` and reformatted all of rust, which seemed wrong. Undoubtedly that used a different toolchain with older style opinions. Thanks.

**Comment by iximeow** [comment]
> How reliably are these tests able to get a nearby mapping to clobber?

This is entirely a question of the OS doing the `mmap`ing. On linuxes I know of, I'm pretty sure mapping just a single page in main should be sufficient, since it'll get the next unreserved page. I've seen `mmap` larger than one page get satisfied with pages later on in memory that don't end up adjacent to the signal stack, which is why the examples have a sequence of single-page mappings.

**Comment by cuviper** [comment]
I raised this to the team on zulip and got no objections, so let's go with it.

@bors r+

**Pull request merged**